[SQL] sqrt(null) causes rust compilation error

A test as such: 

```java
    @Test
    public void testSqrtNull() {
        this.q("""
                SELECT sqrt(null);
                 sqrt
                ------
                null"""
        );
    }
```

Fails to compile to rust: 

```rust
error[E0425]: cannot find function `sqrt_nullN` in this scope
   --> src/lib.rs:167:23
    |
167 |             Tup1::new(sqrt_nullN(None::<()>))
    |                       ^^^^^^^^^^ not found in this scope
```

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -384,6 +384,19 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression> implement
         return call.op.getName().toLowerCase();
     }
 
+    DBSPExpression compilePolymorphicFunction(String opName, CalciteObject node, DBSPType resultType,
+                                              List<DBSPExpression> ops, Integer... expectedArgCount) {
+        this.validateArgCount(node, ops.size(), expectedArgCount);
+        StringBuilder functionName = new StringBuilder(opName);
+        DBSPExpression[] operands = ops.toArray(new DBSPExpression[0]);
+        for (DBSPExpression op: ops) {
+            DBSPType type = op.getType();
+            // Form the function name from the argument types
+            functionName.append("_").append(type.baseTypeWithSuffix());
+        }
+        return new DBSPApplyExpression(node, functionName.toString(), resultType, operands);
+    }
+
     /**
      * Compile a function call into a family of Rust functions,
      * depending on the argument types.
@@ -398,15 +411,7 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression> implement
     DBSPExpression compilePolymorphicFunction(RexCall call, CalciteObject node, DBSPType resultType,
                                     List<DBSPExpression> ops, Integer... expectedArgCount) {
         String opName = this.getCallName(call);
-        this.validateArgCount(node, ops.size(), expectedArgCount);
-        StringBuilder functionName = new StringBuilder(opName);
-        DBSPExpression[] operands = ops.toArray(new DBSPExpression[0]);
-        for (DBSPExpression op: ops) {
-            DBSPType type = op.getType();
-            // Form the function name from the argument types
-            functionName.append("_").append(type.baseTypeWithSuffix());
-        }
-        return new DBSPApplyExpression(node, functionName.toString(), resultType, operands);
+        return this.compilePolymorphicFunction(opName, node, resultType, ops, expectedArgCount);
     }
 
     String typeString(DBSPType type) {
@@ -500,6 +505,12 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression> implement
         DBSPExpression arg = ops.get(argument);
         if (!arg.getType().is(DBSPTypeString.class))
             ops.set(argument, arg.cast(DBSPTypeString.varchar(arg.getType().mayBeNull)));
+    }
+
+    void ensureDouble(List<DBSPExpression> ops, int argument) {
+        DBSPExpression arg = ops.get(argument);
+        if (!arg.getType().is(DBSPTypeDouble.class))
+            ops.set(argument, arg.cast(new DBSPTypeDouble(arg.getType().getNode(), arg.getType().mayBeNull)));
     }
 
     @Override
@@ -688,8 +699,27 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression> implement
                         return this.compilePolymorphicFunction(call, node, type,
                                 ops, 1);
                     }
-                    case "st_distance":
+                    case "st_distance": {
+                        return this.compilePolymorphicFunction(call, node, type,
+                                ops, 2);
+                    }
                     case "power": {
+                        // power(a, .5) -> sqrt(a).  This is more precise.
+                        // Calcite does the opposite conversion.
+                        assert ops.size() == 2: "Expected two arguments for power function";
+                        DBSPExpression argument = ops.get(1);
+                        if (argument.is(DBSPDecimalLiteral.class)) {
+                            DBSPDecimalLiteral dec = argument.to(DBSPDecimalLiteral.class);
+                            BigDecimal pointFive = new BigDecimal(5).movePointLeft(1);
+                            if (!dec.isNull && Objects.requireNonNull(dec.value).equals(pointFive)) {
+                                ops = Linq.list(ops.get(0));
+                                if (ops.get(0).getType().is(DBSPTypeNull.class)) {
+                                    ops.set(0, ops.get(0).cast(type));
+                                }
+                                return this.compilePolymorphicFunction(
+                                        "sqrt", node, type, ops, 1);
+                            }
+                        }
                         return this.compilePolymorphicFunction(call, node, type,
                                 ops, 2);
                     }
@@ -718,18 +748,13 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression> implement
                     case "csc":
                     case "csch":
                     {
-                        // Turn the argument into Double
-                        DBSPExpression exp = ops.get(0);
-                        ops.set(0, exp.cast(new DBSPTypeDouble(node, exp.getType().mayBeNull)));
+                        this.ensureDouble(ops, 0);
                         return this.compilePolymorphicFunction(call, node, type, ops, 1);
                     }
                     case "atan2":
                     {
-                        // Turn the arguments into Double
-                        for (int i = 0; i < ops.size(); i++) {
-                            DBSPExpression arg = ops.get(i);
-                            ops.set(i, arg.cast(new DBSPTypeDouble(node, arg.getType().mayBeNull)));
-                        }
+                        for (int i = 0; i < ops.size(); i++)
+                            this.ensureDouble(ops, i);
                         return this.compilePolymorphicFunction(call, node, type, ops, 2);
                     }
                     case "split":

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/EliminateFunctions.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/EliminateFunctions.java
@@ -3,8 +3,10 @@ package org.dbsp.sqlCompiler.compiler.visitors.inner;
 import org.dbsp.sqlCompiler.compiler.IErrorReporter;
 import org.dbsp.sqlCompiler.compiler.frontend.CalciteObject;
 import org.dbsp.sqlCompiler.compiler.visitors.VisitDecision;
-import org.dbsp.sqlCompiler.ir.expression.*;
-import org.dbsp.sqlCompiler.ir.expression.literal.DBSPDecimalLiteral;
+import org.dbsp.sqlCompiler.ir.expression.DBSPApplyExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPBlockExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPPathExpression;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPStringLiteral;
 import org.dbsp.sqlCompiler.ir.statement.DBSPExpressionStatement;
 import org.dbsp.sqlCompiler.ir.statement.DBSPStatement;
@@ -12,16 +14,13 @@ import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.DBSPTypeTuple;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeVoid;
 
-import java.math.BigDecimal;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Objects;
 import java.util.function.Function;
 
 /*
  * Eliminate some function implementations.
  * For now just:
- * - power(a, .5)
  * - dump(x)
  */
 public class EliminateFunctions extends InnerRewriteVisitor {
@@ -40,20 +39,7 @@ public class EliminateFunctions extends InnerRewriteVisitor {
         DBSPPathExpression path = function.as(DBSPPathExpression.class);
         if (path != null) {
             String functionName = path.path.toString();
-            if (functionName.startsWith("power_")) {
-                // power_base_exp(a, .5) -> sqrt_base(a).
-                String tail = functionName.split("_")[1];
-                assert arguments.length == 2: "Expected two arguments for power function";
-                DBSPExpression argument = arguments[1];
-                if (argument.is(DBSPDecimalLiteral.class)) {
-                    DBSPDecimalLiteral dec = argument.to(DBSPDecimalLiteral.class);
-                    BigDecimal pointFive = new BigDecimal(5).movePointLeft(1);
-                    if (!dec.isNull && Objects.requireNonNull(dec.value).equals(pointFive)) {
-                        String newName = "sqrt_" + tail;
-                        result = new DBSPApplyExpression(node, newName, type, arguments[0]);
-                    }
-                }
-            } else if (functionName.startsWith("dump")) {
+            if (functionName.startsWith("dump")) {
                 // dump(prefix, tuple) -> { print(prefix);
                 //                          print(": (")
                 //                          writelog("%%,", tuple[0]), ...,

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java
@@ -206,6 +206,16 @@ public class FunctionsTest extends SqlIoTest {
     }
 
     @Test
+    public void testSqrtNull() {
+        this.q("""
+                SELECT sqrt(null);
+                 sqrt
+                ------
+                null"""
+        );
+    }
+
+    @Test
     public void testLeftNull() {
         this.q("""
                 SELECT LEFT(NULL, 100);

--- a/sql-to-dbsp-compiler/lib/sqllib/src/lib.rs
+++ b/sql-to-dbsp-compiler/lib/sqllib/src/lib.rs
@@ -806,6 +806,12 @@ pub fn sqrt_decimal(left: Decimal) -> F64 {
 
 some_polymorphic_function1!(sqrt, decimal, Decimal, F64);
 
+pub fn sqrt_d(left: F64) -> F64 {
+    F64::from(left.into_inner().sqrt())
+}
+
+some_polymorphic_function1!(sqrt, d, F64, F64);
+
 //////////////////// floor /////////////////////
 
 #[inline(always)]

